agent/src/util/device_plugin_builder.rs

```diff
@@ -0,0 +1,202 @@
+use super::{
+    constants::{K8S_DEVICE_PLUGIN_VERSION, KUBELET_SOCKET},
+    device_plugin_service::{DevicePluginService, InstanceMap},
+    v1beta1,
+    v1beta1::{device_plugin_server::DevicePluginServer, registration_client, DevicePluginOptions},
+};
+use akri_discovery_utils::discovery::v0::Device;
+use akri_shared::{
+    akri::{configuration::KubeAkriConfig, AKRI_PREFIX},
+    uds::unix_stream,
+};
+use async_trait::async_trait;
+use futures::stream::TryStreamExt;
+use log::{info, trace};
+#[cfg(test)]
+use mockall::{automock, predicate::*};
+use std::{convert::TryFrom, env, path::Path, time::SystemTime};
+use tokio::{
+    net::UnixListener,
+    net::UnixStream,
+    sync::{broadcast, mpsc},
+    task,
+};
+use tonic::transport::{Endpoint, Server, Uri};
+use tower::service_fn;
+
+#[cfg_attr(test, automock)]
+#[async_trait]
+pub trait DevicePluginBuilderInterface: Send + Sync {
+    async fn build_device_plugin(
+        &self,
+        instance_name: String,
+        config: &KubeAkriConfig,
+        shared: bool,
+        instance_map: InstanceMap,
+        device_plugin_path: &str,
+        device: Device,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>>;
+}
+
+/// For each Instance, builds a Device Plugin, registers it with the kubelet, and serves it over UDS.
+pub struct DevicePluginBuilder {}
+
+#[async_trait]
+impl DevicePluginBuilderInterface for DevicePluginBuilder {
+    /// This creates a new DevicePluginService for an instance and registers it with the kubelet
+    async fn build_device_plugin(
+        &self,
+        instance_name: String,
+        config: &KubeAkriConfig,
+        shared: bool,
+        instance_map: InstanceMap,
+        device_plugin_path: &str,
+        device: Device,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        info!("build_device_plugin - entered for device {}", instance_name);
+        let capability_id: String = format!("{}/{}", AKRI_PREFIX, instance_name);
+        let unique_time = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH)?;
+        let device_endpoint: String = format!("{}-{}.sock", instance_name, unique_time.as_secs());
+        let socket_path: String = Path::new(device_plugin_path)
+            .join(device_endpoint.clone())
+            .to_str()
+            .unwrap()
+            .to_string();
+        // Channel capacity set to 6 because 3 possible senders (allocate, update_connectivity_status, and handle_config_delete)
+        // and and receiver only periodically checks channel
```

Comment:
nit: typo and and

Reply:
Also, why 6 when there are 3 senders?

Reply:
I did 6 so as to double in the case they all send twice before LAW read it. 
I am not sure what the best capacity is here either. One of the senders is `update_connectivity_status`, which tells list and watch whenever devices go offline or online so it can then tell kubelet that the "virtual devices" are UNHEALTHY/HEALTHY.  If a device keeps goes offline/online/offline/online faster than list_and_watch can build a list and watch response, the channel would get clogged. This seems unlikely though.

Reply:
add inline documentation in fa4eba408f57f2